internal/iamauth/responses/arn.go

```diff
@@ -0,0 +1,94 @@
+package responses
+
+import (
+	"fmt"
+	"strings"
+)
+
+// https://github.com/hashicorp/vault/blob/ba533d006f2244103648785ebfe8a9a9763d2b6e/builtin/credential/aws/path_login.go#L1722-L1744
+type ParsedArn struct {
+	Partition     string
+	AccountNumber string
+	Type          string
+	Path          string
+	FriendlyName  string
+	SessionInfo   string
+}
+
+// https://github.com/hashicorp/vault/blob/ba533d006f2244103648785ebfe8a9a9763d2b6e/builtin/credential/aws/path_login.go#L1482-L1530
+// However, instance profiles are not support in Consul.
+func ParseArn(iamArn string) (*ParsedArn, error) {
```

Comment:
Compared to the Vault implementation of this, I added a few checks here for empty strings on parts, and it doesn't support ARNs for EC2 instance profiles, since those are only supported by Vault for the EC2 auth method.

(This could be simplified a bit by using [arn.ParseArn](https://docs.aws.amazon.com/sdk-for-go/api/aws/arn/#Parse))